### DISCUSSION TITLE
Is Milvus The Right Choice?

### DISCUSSION THREAD

**Discussion by Kencho5** [milvus-io/milvus#15203] [Q&A and General discussion]
hello,

i want to know if milvus is the right thing to choose for my project. i’ve got around 160 million images (each image is 32x32). i want to insert those images in milvus so i want to know if it’s possible to insert those images in a fast and efficient way, without it taking more than 2-3 days. also i want to know it i will be able to search those images from input. for example i’ve got an image that’s slightly different than the original one in the milvus database, each pixel is off by 1-2 rgb value so think of it as a slight tint on it. will i be able to search similar images from there? 

thanks.

**Comment by yhmo** [comment]
Milvus doesn't store image files, it only stores vectors/embeddings. Typically users use a SQL db to store the mapping relationship between vectors/embeddings and images.

The recommended solution for your requirement:
  1. Generate vectors/embeddings from your images
We have a project named "[Towhee](https://towhee.io/)" to do this work. It can convert images/audios into vectors/embeddings within 10 lines script.
Sounds your requirement is similar to [this pipeline](https://docs.towhee.io/tutorials/image-deduplication). Just 3 lines to generate an embedding for an image:
```
from towhee import pipeline
embedding_pipeline = pipeline('image-encoding')
img0_embedding = embedding_pipeline('/path/to/img0')
```
To install this tool, follow [this page](https://docs.towhee.io/get-started/install).

  2. Insert the vectors/embeddings into Milvus, build a mapping relationship between embeddings and image paths
Now you can write a script to do this work:
use Towhee to generate embeddings for your images
insert embeddings into Milvus, we recommend inserting batch by batch. For example, generate 100 embeddings, insert the 100 embeddings, then generate the next 100 embeddings, insert ...
for each insert API call, the Milvus server will return an id array of the inserted embeddings(you also can specify one id for each embedding by yourself), use a SQL database to store the id and image path relationship.

  3. Search similar vectors
To search an image, you call the Towhee to generate an embedding again, then use the embedding to search in Milvus. Milvue returns a top result, the top one is the most similar image's id.

  4. Find out the image by the mapping relationship
You use the id to query in the SQL database, finally, you get the image path.

**Reply by Kencho5** [reply]
so i need to convert images into vectors, insert vectors in milvus, insert id returned from milvus into a sql database. will this be fast enough to insert 170 million images though? and would i need to store images on a disk? or i just need vectors

**Reply by Kencho5** [reply]
also, do i need towhee for that? can i normally transfer an image to vector array and store that in milvus?

**Reply by yhmo** [reply]
The key point is to build a mapping relationship between the vectors and your images. If your images are http url, then build am id-url mapping. If your images are stored on disk, then build an id-path mapping.
You can convert images to vectors with your tools, any tool is ok. 

Since 170 million images is a huge number, it would be very time-consuming work.

**Reply by Kencho5** [reply]
so basically there is no fast way to transform 170 images into vectors and insert them in milvus without it taking weeks of running

**Reply by yhmo** [reply]
I think the bottleneck is in "transform 170 images into vectors",  by using resnet50, it cost about 50~100ms for each image.
If you insert vectors into milvus by batch, each second you can insert thousands of vectors or even ten thousand vectors.

**Reply by Kencho5** [reply]
i’ve got image urls not the images itself. i can grab image bytes from thousands of images at once using async though. would that work?

**Reply by Kencho5** [reply]
can't towhee run in multiprocessing using the **map()** function? i can't seem to get it to run

**Comment by Kencho5** [comment]
i seem to have trouble running **embedding_pipeline()** function in parallel/async. when i tried with multiprocessing and async task gathering both waited for one thread to finish and after started another one to vector the images. is it possible to run it in parallel?

**Reply by yhmo** [reply]
I saw your [issue](https://github.com/towhee-io/towhee/issues/480) in the towhee project. Seems it is a problem, towhee team will fix it later.
